pkg/verify/verify.go

```diff
@@ -0,0 +1,198 @@
+	"github.com/sigstore/rekor/pkg/generated/client"
+	"github.com/sigstore/rekor/pkg/generated/client/tlog"
+	"github.com/sigstore/rekor/pkg/generated/models"
+	"github.com/sigstore/rekor/pkg/util"
+	"github.com/sigstore/sigstore/pkg/signature"
+	"github.com/sigstore/sigstore/pkg/signature/options"
+	"github.com/transparency-dev/merkle/proof"
+	"github.com/transparency-dev/merkle/rfc6962"
+)
+
+// ProveConsistency verifies consistency between an initial, trusted STH
+// and a second new STH. Assumes that the signature on the STHs' are verified.
+func ProveConsistency(ctx context.Context, rClient *client.Rekor,
+	oldSTH *util.SignedCheckpoint, newSTH *util.SignedCheckpoint, treeID string) error {
+	oldTreeSize := int64(oldSTH.Size)
+	switch {
+	case oldTreeSize == int64(newSTH.Size):
+		if !bytes.Equal(oldSTH.Hash, newSTH.Hash) {
+			return errors.New("old root hash does not match STH hash")
+		}
+	case oldTreeSize < int64(newSTH.Size):
+		consistencyParams := tlog.NewGetLogProofParamsWithContext(ctx)
+		consistencyParams.FirstSize = &oldTreeSize      // Root size at the old, or trusted state.
+		consistencyParams.LastSize = int64(newSTH.Size) // Root size at the new state to verify against.
+		consistencyParams.TreeID = &treeID
+		consistencyProof, err := rClient.Tlog.GetLogProof(consistencyParams)
+		if err != nil {
+			return err
+		}
+		var hashes [][]byte
+		for _, h := range consistencyProof.Payload.Hashes {
+			b, err := hex.DecodeString(h)
+			if err != nil {
+				return errors.New("error decoding consistency proof hashes")
+			}
+			hashes = append(hashes, b)
+		}
+		if err := proof.VerifyConsistency(rfc6962.DefaultHasher,
+			oldSTH.Size, newSTH.Size, hashes, oldSTH.Hash, newSTH.Hash); err != nil {
+			return err
+		}
+	case oldTreeSize > int64(newSTH.Size):
+		return errors.New("inclusion proof returned a tree size larger than the verified tree size")
+	}
+	return nil
+
+}
+
+// VerifyCurrentCheckpoint verifies the provided checkpoint by verifying consistency
+// against a newly fetched Checkpoint.
+//nolint
+func VerifyCurrentCheckpoint(ctx context.Context, rClient *client.Rekor, verifier signature.Verifier,
+	oldSTH *util.SignedCheckpoint, treeID string) error {
+	// Get and verify against the current STH.
+	infoParams := tlog.NewGetLogInfoParamsWithContext(ctx)
+	result, err := rClient.Tlog.GetLogInfo(infoParams)
+	if err != nil {
+		return err
+	}
+
+	logInfo := result.GetPayload()
+	sth := util.SignedCheckpoint{}
+	if err := sth.UnmarshalText([]byte(*logInfo.SignedTreeHead)); err != nil {
+		return err
+	}
+
+	// Verify the signature on the SignedCheckpoint.
+	if !sth.Verify(verifier) {
+		return errors.New("signature on tree head did not verify")
+	}
+
+	// Now verify consistency up to the STH.
+	return ProveConsistency(ctx, rClient, oldSTH, &sth, treeID)
+}
+
+// VerifyInclusion verifies an entry's inclusion proof. Clients MUST also
+// verify the root hash via VerifyRootHash.
```

Comment:
nit: VerifyRootHash is the old function name

Nonetheless, a consistency proof is effectively an inclusion proof for each checkpoint, correct? So just checking consistency would be sufficient if you have a previous checkpoint

Reply:
I'm confused about the statement "consistency proof is effectively an inclusion proof for each checkpoint".

I'll update the comment 
```
 Clients MUST either verify
// the root hash against a new STH (via VerifyCurrentCheckpoint) or against a
// trusted, existing STH (via ProveConsistency).
```

Reply:
Chatted offline, I was incorrect cause I was missing the context on a comment on a doc. 

The comment SG. Should we note that if you verify against the current checkpoint, the inclusion proof should have been one that's been persisted and is trusted?

Reply:
against the current checkpoint? then wouldn't you persist the new one? 

i think that's usage / client storage is out of scope here

Reply:
I was asking when you'd use VerifyInclusion then VerifyCurrentCheckpoint. What is the use case? You're right that you'd just persist the new one, so I guess my point is that you'd probably only call ProveConsistency after VerifyInclusion

Reply:
Like the above comment, I think we should document the examples for various use cases to make it clear when these are to be used.

* A witness verifying from a previous checkpoint - VerifyCurrentCheckpoint
* Verifying an SET - VerifyLogEntry
* Verifying an inclusion proof - (I think) VerifyInclusion and ProveConsistency
* Verifying an inclusion proof offline - Just VerifyInclusion (and verifying STH signature, which isn't a function currently)

Reply:
I think this should be in a documentation file. Documentation in code would only be understood anyway from code examples.